create_file function and unit tests for it

alicjakwie opened the pull request.

alicjakwie committed:
Created create_file.go and create_file_test.go

alicjakwie committed:
Created create_file.go which integrates previous file_gen and the CreateFile function

evanSpendlove commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (f *RandomFile) NewReader() *randomFileReader {
+	return &randomFileReader{size: f.Size, rand: rand.New(rand.NewSource(f.Seed))}
```
You're going to want to use a lockedSource to make this thread-safe. Sources aren't thread-safe by default.

```
type lockedSource struct {
	lk  sync.Mutex
	src *rngSource
}
```

See: https://golang.org/src/math/rand/rand.go

leczb replied:
It's unlikely that a Reader instances would be used in multiple threads.
Also note that only Hermes is going to use this: just add a comment stating that it's not thread safe.

evanSpendlove replied:
If the reader itself has a rand variable with a source, can multiple readers exist with different sources without conflicting? If so, then there's no issue with this.

Hermes will be using an instance of this Reader in each probe iteration, so there would be multiple in existence if multiple targets are being monitored.

leczb replied:
NewReader() returns a new instance every time. Each instance is independent.
Each instance is going to be used for a short period for a single purpose: either to calculate the checksum or to send the contents to the storage system.
I don't expect these instances to be retained, so no locking is necessary.

evanSpendlove commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,116 @@
+type randomFileReader struct {
+	size int
+	// Current reading index.
+	i    int
```
Could you change this to `idx` or something slightly more descriptive, as this is very terse.

leczb replied:
It's OK to have short variable names for such small-scoped variables (it's only used in the next ~10 lines).
Also, `i` is generally used as the index variable.
The standard library can be used as a reference:
https://golang.org/src/strings/reader.go

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	r := f.NewReader()
+	h := sha1.New()
+	buff := make([]byte, 8)
+	for {
```
There's no need for this loop and the buffer.
io.Copy() takes care of everything.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// returns checksum in hex notation
+	return fmt.Sprintf("%x", h.Sum(nil)), nil
```
Don't convert to a string here.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	if err != nil {
+		return nil, err
+	}
+	return fmt.Sprintf("Hermes_%02d_%v", fileID, checksum), nil
```
...convert it here instead.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+		return fmt.Errorf("At %v the file ID provided wasn't in the required range [0,50]", time.Now())
+	}
+	f := RandomFile{Seed: fileID, Size: fileSize}
+	fileName, err := f.generateFileName(fileID)
```
It's surprising that f.generateFileName() requires the file ID.
I think it would be cleaner if RandomFile had methods like FileName() and CheckSum() that returned these values.

alicjakwie replied:
changed!

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	r := f.NewReader()
+
+	wc := client.Bucket(bucketName).Object(fileName).NewWriter(ctx)
+	buff := make([]byte, 8)
```
Again, there's no need for a buffer and a loop.
Also note that an 8 byte buffer would be too small, making this loop unnecessarily slow.

cianbgoog commented on probe/create_file.go:
```diff
@@ -0,0 +1,4 @@
+package probe
```
Why is this file here?

alicjakwie replied:
it's the create file function (part of the monitor probe)

deniadelcid replied:
Can't the probe monitor just call the `CreateFile` function from hermes/probe/create/create_file.go instead?

leczb replied:
Remove empty files from the PR.

cianbgoog commented on probe/create_file_test.go:
```diff
@@ -0,0 +1,5 @@
+package probe
```
Why is this file here?

cianbgoog commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int, fileSize int) error {
+	if fileID < 1 || fileID > 50 {
```
We use these same numbers in delete file - could we merge this into a constant somewhere?

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+		if _, err = io.Copy(h, r); err != nil {
+			return nil, fmt.Errorf("io.Copy: %v", err)
+		}
+
```
```suggestion
```

cianbgoog commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+		return fmt.Errorf("storage.NewClient: %v", err)
+	}
+	defer client.Close()
+	ctx, cancel := context.WithTimeout(ctx, p.opts.Timeout)
```
Do we need to `cancel` this? Or could we just not cancel it, and use the existing `context`?

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+type randomFileReader struct {
+	size int
+	// Current reading index.
+	i    int
```
nit: I would move this comment inline
```suggestion
	i    int  // Current reading index.
```

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+		n, err = r.rand.Read(b)
+	}
+	r.i += n
+	return
```
Although this works, can you add return values here, to make it consistent with the other functions?

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	} else {
+		n, err = r.rand.Read(b)
+	}
+	r.i += n
```
- [] You are collecting the errors from the `r.rand.Read`, but you aren't handling it, is that intended?

alicjakwie replied:
Thank you for pointing this out this was not intended!

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+
+type randomFileReader struct {
+	size int
+	// Current reading index.
```
Move this comment inline

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+		n, err = r.rand.Read(b[:r.size-r.i])
+	} else {
+		n, err = r.rand.Read(b)
+	}
```
I think you could get rid of this `else` by doing something like this:

```
buf := b
if len(b) > r.size-r.i {
		buf = b[:r.size-r.i]
}
n, err = r.rand.Read(buf)
```

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	r := f.NewReader()
+
+	wc := client.Bucket(bucketName).Object(fileName).NewWriter(ctx)
+	buff := make([]byte, 8)
```
Why is this 8 bytes?

deniadelcid replied:
Maybe you want to make this a flag or a constant (and provide some reasoning to its size).

alicjakwie replied:
buff was deleted and is no longer part of the code. now it's just io.Copy(r, wc) where r is the reader of a RandomHermesFile

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+		}
+
+	}
+	//io
```
Either get rid of this comment of add a more meaningful comment that isn't covered by the context.

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,114 @@
+	return fmt.Sprintf("Hermes_%02d_%v", fileID, checksum), nil
+}
+
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int, fileSize int) error {
```
-[] Since CreateFile doesn't return the file its created, You could add a description at here, at the top explaining where it's stored. 

-[] Also, since this is an exported method, add a comment at the top where you explain what each of your parameters/arguments are and any quirks they might have. Similarly add a section for their returns where you explain every error message or success scenario would look like and what they represent. (just like you did in file_gen.go that is in the master branch)

alicjakwie replied:
doc strings added

deniadelcid reviewed (commented):
- [x] I'm just a little confused, why do we need two different create_file.go files in two underlying probe directories. Can these be merged into a single one and if not what is the purpose of having them separate?

- [x] Can you run gofmt on this file?

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	"fmt"
+	"io"
+	"math/rand"
+	// "os"
```
Remove commented code.
In general: don't commit commented code.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+const (
+	begin = 1  // ID of the first HermesFile
+	end   = 50 // ID of the last HermesFile
+	maxFileSize = 1000 // maximum allowed file size in bytes 
```
It's best to include the units in the name to remove ambiguity: `maxFileSizeBytes`.

alicjakwie replied:
noted!

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	maxFileSize = 1000 // maximum allowed file size in bytes 
+)
+
+type RandomHermesFile struct {
```
I don't think it's necessary to include "Hermes" in the name.

alicjakwie replied:
Removed

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	rand *rand.Rand
+}
+
+func (r *randomHermesFileReader) readDone() bool {
```
I think moving these conditions into methods is unnecessary complication and makes the reader jump back-and-forth. I would keep them in-line.

alicjakwie replied:
It helped me a lot with 'studying' the code esp when debugging if that is alright I will leave that in when working on the code and will add a todo to move them back to inline?

alicjakwie replied:
If not I am sorry and I will remove them now

leczb replied:
You are right that moving tricky condition expressions into well-named functions makes the code more readable.
However, this is rather simple code and in this case it makes it a bit harder to read as the reader has to jump back-and-forth. If you think the conditions don't speak for themselves, adding a clarifying comments would be better in this case.

alicjakwie replied:
moved to inline and added clarifying comments

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	return len(buffer) > r.bytesLeft()
+}
+
+func (r *randomHermesFileReader) Read(buff []byte) (n int, err error) {
```
Nit: the idiomatic variable name for a buffer is `buf` or just `b`.

alicjakwie replied:
changed

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+		b = buff[:r.bytesLeft()]
+	}
+	n, err = r.rand.Read(b)
+	if(err!=nil){
```
Nit: remove the parentheses.
Also: make sure you run `gofmt` before sending the code for review.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	}
+	n, err = r.rand.Read(b)
+	if(err!=nil){
+		return 0, err // 0 as nil casn't be returned as a type int argument
```
There's no need for this comment. It's a well-established idiom to return the "zero value" of the type. (nil is just the zero value for pointer types).

alicjakwie replied:
removed

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+
+// Warning: NewReader is not thread safe
+func (f *RandomHermesFile) NewReader() *randomHermesFileReader {
+	return &randomHermesFileReader{size: f.Size, rand: rand.New(rand.NewSource(f.ID))} // ID will serve as a Seed and i will be set to 0 automatically
```
Nit: move each field to its own line. After that, move the comment above the line with NewSource().

alicjakwie replied:
moved

cianbgoog commented on hermes/probe/create_file.go:
```diff
@@ -6,109 +6,119 @@ import (
+	//defer cancel()
 	r := f.NewReader()
-
+	fmt.Println("Does everything work up until bucket naming?")
```
probs want to remove this (and similar fmt.Println lines) :-)

alicjakwie replied:
noted and removed! They were in the old version for debugging purposes

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	if _, err := io.Copy(h, r); err != nil {
+		return nil, fmt.Errorf("io.Copy: %v", err)
+	}
+	// returns checksum in hex notation
```
Remove the comment (it's also not true anymore).

alicjakwie replied:
removed

cianbgoog commented on hermes/probe/create_file.go:
```diff
@@ -6,109 +6,119 @@ import (
-	ctx, cancel := context.WithTimeout(ctx, p.opts.Timeout)
-	defer cancel()
+	defer client.Close()*/
+	//ctx, cancel := context.WithTimeout(ctx, p.opts.Timeout)
```
Remove this, rather than commenting it out (here & elsewhere).

alicjakwie replied:
noted! git replaced with and committed an old version not ready for a review

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func (f *RandomHermesFile) FileName() (string, error) {
+	if f.ID < begin || f.ID > end {
```
I think it would be cleaner to do this in a "constructor" function like this: `func NewRandomFile(id int) (*RandomFile, error)`.

alicjakwie replied:
moved the code to a constructor

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+
+func (f *RandomHermesFile) FileName() (string, error) {
+	if f.ID < begin || f.ID > end {
+		return "", fmt.Errorf("The file ID provided %v wasn't in the required range [1,50]", f.ID)
```
Nit: error messages should start with lower-case (as opposed to log messages, which start with upper-case).

alicjakwie replied:
noted changed! Left upper case letter only when the message started with an exported function name

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+
+func (f *RandomHermesFile) FileName() (string, error) {
+	if f.ID < begin || f.ID > end {
+		return "", fmt.Errorf("The file ID provided %v wasn't in the required range [1,50]", f.ID)
```
Nit: There are consts for the lower and upper bound. Use those instead of hard-coding the values in the string.
Otherwise it's very hard to keep the in sync.

alicjakwie replied:
changed

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+		return "", fmt.Errorf("The file ID provided %v wasn't in the required range [1,50]", f.ID)
+	}
+	if f.Size > maxFileSize {
+		return "", fmt.Errorf("The file size provided %v bytes exceeded the limit %v bytes", f.Size, maxFileSize)
```
The idiomatic error message would be `"invalid argument: RandomFile.Size = %d; want 0 < size <= %d", f.Size, maxFileSize`.

alicjakwie replied:
Changed the error message

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("Hermes_%02d_%v", f.ID, fmt.Sprintf("%x", checksum)), nil
```
Skip the second Sprintf:
```
return fmt.Sprintf("Hermes_%02d_%x", f.ID, checksum), nil
```

alicjakwie replied:
done

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	return fmt.Sprintf("Hermes_%02d_%v", f.ID, fmt.Sprintf("%x", checksum)), nil
+}
+
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int64, fileSize int, client stiface.Client) error {
```
Nit: Consecutive arguments sharing the same type can be listed as `foo, bar, baz string`.

leczb replied:
Nit: Make the naming consistent: either use `bucket` or `bucketName` everywhere.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	if err != nil {
+		return err
+	}
+	/*client, err := storage.NewClient(ctx)
```
Remove commented code.

alicjakwie replied:
removed

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int64, fileSize int, client stiface.Client) error {
+	f := RandomHermesFile{ID: fileID, Size: fileSize}
+	fileName, err := f.FileName()
+	fmt.Println("Got Here File successfully named")
```
Remove debugging code.

cianbgoog commented on hermes/probe/create_file_test.go:
```diff
@@ -1 +1,201 @@
+)
+
+func TestFileName(t *testing.T) {
+	testUnits := [6]RandomHermesFile{ 
```
Rather than writing this multiple times, it's normal to do something like 
```
tests := []struct{file RandomHermesFile
                            wantName string
                             wantError error}{
                              {RandomHermesFile{51, 12}, "",
                               "The file ID provided 51 wasn't in the required range [1, 50]"}
                              }
for _, test := range tests {
  got, err := test.file.FileName()
  if got != test.wantName {
    t.Errorf("descriptive error message: HermesFile{%v}.FileName(): wanted: %v got: %v", test.file, test.wantName, test.wantError)
}
<test for the error message goes here>
```

This makes the tests an awful lot easier to read.

Apologies for typos in my code example, github comment boxes are not a great editor :-P

This, here & for other tests.

alicjakwie replied:
Changed! Wayyyyy more readable now thank you!

cianbgoog commented on hermes/probe/create_file_test.go:
```diff
@@ -1 +1,201 @@
+	}
+
+}
+
```
This stiface is probably best put in another file, since I assume it'll be used by all your tests?

alicjakwie replied:
moved

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	return &randomHermesFileReader{size: f.Size, rand: rand.New(rand.NewSource(f.ID))} // ID will serve as a Seed and i will be set to 0 automatically
+}
+
+func (f *RandomHermesFile) CheckSum() ([]byte, error) {
```
Is it expected that the checksum will only be calculated once during the lifetime of the object?
If it's needed more than once, it would better to store the result instead of re-calculating it.

alicjakwie replied:
The checksum will be calculated once when the object is created and stored in the string form in the fileName.
It will then in readFile be calculated again and compared with the one stored in the string form in the fileName to make sure that the contents of the file haven't changed

leczb replied:
Do you mean the checksum will be recovered from the filename by parsing the string?
I would advise against that.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	//ctx, cancel := context.WithTimeout(ctx, p.opts.Timeout)
+	//defer cancel()
+	r := f.NewReader()
+	fmt.Println("Does everything work up until bucket naming?")
```
Remove debugging code.

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+	r := f.NewReader()
+	fmt.Println("Does everything work up until bucket naming?")
+	wc := client.Bucket(bucketName).Object(fileName).NewWriter(ctx)
+	fmt.Println("Storage reader successfully created")
```
Remove debugging code.

alicjakwie replied:
removed

leczb commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,124 @@
+		return fmt.Errorf("Writer.Close: %v", err)
+	}
+	return nil
+
```
Nit: Remove empty line.

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+		{3, 0}, // Invalid Size
+		{3, 1001}, // Invalid Size
+	}
+	// Case: testUnits[0] RandomHermesFile{51, 12}
```
Use table-driven tests instead.

alicjakwie replied:
changed to table-driven tests

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+)
+
+func TestFileName(t *testing.T) {
+	testUnits := [6]RandomHermesFile{ 
```
Nit: Use slices instead of arrays (that lets you omit the size).

alicjakwie replied:
changed

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+
+func TestFileName(t *testing.T) {
+	testUnits := [6]RandomHermesFile{ 
+		{51, 12}, // Invalid ID  
```
Put the test case description in string `desc` (or `name` or `description`) field instead of a comment.

alicjakwie replied:
no longer applies

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+
+}
+
+type fakeBucket struct {
```
I would move the fake implementation into its own package, as I expect it will be used in all tests.

alicjakwie replied:
moved

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+	bucketName := "test_bucket_probe0"
+	client := newFakeClient()
+	fbh := client.Bucket(bucketName) // fakeBucketHandle
+	err := fbh.Create(ctx, bucketName, nil) // creates the bucket with name "test_bucket_probe0"
```
Check the error here like this: `if err := ...; err != nil {`

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+	fbh := client.Bucket(bucketName) // fakeBucketHandle
+	err := fbh.Create(ctx, bucketName, nil) // creates the bucket with name "test_bucket_probe0"
+	mp := &MonitorProbe{}
+	var fileID int64 = 6
```
Use `:=` instead of `var`.

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+	var fileID int64 = 6
+	fileSize := 50
+	target := "ossdawdler-step2020"
+	err = mp.CreateFile(ctx, target, bucketName, fileID, fileSize, client)
```
`if err := ...; err != nil {`

leczb commented on hermes/probe/create_file_test.go:
```diff
@@ -0,0 +1,201 @@
+	if(err!=nil){
+		t.Error(err)
+	}
+}
```
Nit: missing newline.

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -6,109 +6,119 @@ import (
-	}
-	f := RandomFile{Seed: fileID, Size: fileSize}
-	fileName, err := f.generateFileName(fileID)
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int64, fileSize int, client stiface.Client) error {
```
- [ ] Would it make more sense to make these a methods of Targets instead? So that  a MonitorProbe (who has a field `targets` of  type `[]*Target` can make calls such as `t.CreateFile()`. Plus you don't appear to use "p" any more
- [  ] `target` shouldn't be a string now that there is a `Target` struct in monitor.go.
- [ ] Make client a pointer instead
- [ ] If you need the logger of a MonitorProbe, you can pass that as a pointer within the arguments.

deniadelcid replied:
Oh, if you keep `p`, rename it to `mp` just to keep it consistent with the rest.

alicjakwie replied:
p removed, CreateFile is now a function and it takes in Targets as an argument as they will be defined in a separate file and maybe even package
Client was changed to a pointer but now will be changed back, because as @leczb pointed out interfaces are stored as a tuple (with first parameter pointer and the second one can be nil so they are in themselves pointers)

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -6,109 +6,119 @@ import (
-			return fmt.Errorf("io.Copy: %v", err)
-		}
-
+	fmt.Println("Storage reader successfully created")
```
You may want to log this message instead to the MonitorProbe's logger instead.

alicjakwie replied:
This was now removed as the old version of the file has been uploaded - it was printed for debugging purposes

deniadelcid commented on hermes/probe/create_file_test.go:
```diff
@@ -1 +1,201 @@
+}
+
+func TestChecksum (t *testing.T) {
+	f := RandomHermesFile{11, 100}
+	fTwo := RandomHermesFile{13, 1000}
+	fCopy := RandomHermesFile{11, 100}
```
You could use a table driven test [1] to define these instead (and as other have pointed out, it's also applicable in your other test functions.

[1] https://github.com/golang/go/wiki/TableDrivenTests

deniadelcid commented on hermes/probe/create_file_test.go:
```diff
@@ -1 +1,201 @@
+	if(err!=nil){
+		t.Error(err)
+	}
+	if fmt.Sprintf("%x",checksum) != fmt.Sprintf("%x",checksumCopy) { //comparing checksums converted to strings as a slice can only be compared to nil
+		t.Errorf("Checksum returned different values for the same RandomHermesFiles {%v, %v}", f.ID, f.Size)
+	}
```
When you create a table to drive this test, you can still make this check by modifying your test struct to have two or three `RandomHermesFile` fields one for the main file, another one for the following (and/or preceding file), and then one for the copy of the main file. You could also just store the actual, expected checksum of the main file as `want`.

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -6,109 +6,119 @@ import (
-	}
-	f := RandomFile{Seed: fileID, Size: fileSize}
-	fileName, err := f.generateFileName(fileID)
+func (p *MonitorProbe) CreateFile(ctx context.Context, target string, bucketName string, fileID int64, fileSize int, client stiface.Client) error {
```
Should CreateFile also support the creation of the NIL file?

alicjakwie replied:
No, it shall not :) there will be a function (called in probe.go) that overwrites the NIL file after each state changing operation

deniadelcid commented on hermes/probe/create_file_test.go:
```diff
@@ -1 +1,201 @@
+	mp := &MonitorProbe{}
+	var fileID int64 = 6
+	fileSize := 50
+	target := "ossdawdler-step2020"
```
Does this have to be the same as a GCP project? If not, we can choose something more generic instead for test purposes.

alicjakwie replied:
tagret string has now been removed

alicjakwie committed:
Created create_file.go and create_file_test.go create_file_test  mocks
GCS and the CreateFile function takes the right arguments

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,123 @@
+   }
+   n, err = r.rand.Read(b)  //n is the length  of the buffer
+   if(err != nil){
+       return  n, err// 0 as nil casn't be returned as a type int argument
```
nit: leave two spaces before the comment.
```suggestion
       return  n, err  // 0 as nil can't be returned as a type int argument
```

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,123 @@
+ 
+// Warning: NewReader is not thread safe
+func (f *RandomHermesFile) NewReader() *randomHermesFileReader {
+   return &randomHermesFileReader{size: f.Size, rand: rand.New(rand.NewSource(f.ID))} // ID will serve as a Seed and i will be set to 0 automatically
```
nit: leave two space before inline comments.

It's unclear what `i` is in this function, can you rephrase the comment? Or, instead, you could just comment right next to where this `i` is defined that it defaults to 0.

alicjakwie replied:
comment rephrased - hope it's more readable now!

deniadelcid commented on hermes/probe/create_file.go:
```diff
@@ -0,0 +1,123 @@
+   }
+   return nil
+ 
+}
```
I think you need a new line here. (Just type an enter)

alicjakwie committed:
Updated CreateFile added metrics restructured packages

alicjakwie committed:
Updated create_test_file.go restructured packages

moved the code responsible for mocking gcs to a different file (mock_gcs.go)
Restructured packages
Implemented table driven tests for  TestFileName
All tests pass

alicjakwie committed:
Rename hermes/probe/create_file_test.go to hermes/probe/create/create_file_test.go

alicjakwie committed:
Update create_file.go

evanSpendlove commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+		}
+		target.LatencyMetrics.APICallLatency[m.APICreateFile][status].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+		return fmt.Errorf("CreateFile(id: %d).%v: could not create file %s: %w", fileID, status, fileName, err)
+	}
```
Currently, metrics are only recorded if the operation fails. Please add a line recording the metric when it passes.

alicjakwie replied:
added

evanSpendlove commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+		var status m.ExitStatus
+		status = m.UnknownFileFound
+		return fmt.Errorf("CreateFile(ID: %d).%v could not create file as file with this ID already exists", fileID, status)
+	}
```
Add the intent of the operation with the filename to the Journal here please.

alicjakwie replied:
added

evanSpendlove commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	target.Journal.Filenames[fileID] = fileName
+	if logger != nil {
+		logger.Infof("Object %v added in bucket %s.", fileName, bucketName)
+	}
```
Add a check here to see if the file was actually created on the file system.

alicjakwie replied:
added

alicjakwie committed:
Update create_file.go

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+			if got != "" {
+				t.Errorf("{%v, %v}.FileName() failed expected empty string, got %s", test.file.ID, test.file.Size, got)
+			}
+			if err.Error() != test.wantErr {
```
Requiring a concrete error message makes the tests brittle. These kinds of tests are called "change detection tests", which detect that the code was touched. It requires busy-work: any time you change the code, the tests have to be updated as well.
It's better to just check whether an error was returned. If the caller needs to distinguish between different error types, then also check that the correct error type is returned. Otherwise, just check whether `err` is nil/not-nil, as required by the test case. Just change the type of `wantErr` to `bool` to indicate whether an error is expected.

alicjakwie replied:
Changed wantErr to bool

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	}
+	n, err = r.rand.Read(b) //n is the length  of the buffer
+	if err != nil {
+		return n, err // in this case n=0 and we return 0 as nil can't be returned as a type int argument
```
Remember to leave two spaces before inline comments and one space after the `//`. In general this would look like: 
` <code><space><space>//<space>your comment`

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	maxFileSizeBytes = 1000 // maximum allowed file size in bytes
+)
+
+type RandomHermesFile struct {
```
For this and the rest of the file, it's unnecessary to include "Hermes" in the name.

alicjakwie replied:
removed

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	return (r.size - r.i)
+}
+
+func (r *randomHermesFileReader) bufferTooLong(buffer []byte) bool {
+	return len(buffer) > r.bytesLeft()
+}
```
To make this consistent with the rest of the file, rename the instances of `buffer` to just `buf`.

alicjakwie replied:
done

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	return n, err
+}
+
+// Warning: NewReader is not thread safe
+func (f *RandomHermesFile) NewReader() *randomHermesFileReader {
```
Either add a doc string or a TODO to add the doc strings. (Here and in the other exported method/functions).

alicjakwie replied:
Created an Issue and added a TODO

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+
+// Warning: NewReader is not thread safe
+func (f *RandomHermesFile) NewReader() *randomHermesFileReader {
+	return &randomHermesFileReader{size: f.Size, rand: rand.New(rand.NewSource(f.ID))} // ID will serve as a Seed and i - index of the currently read byte  will be set to 0 automatically in the returned reader
```
nit: The comment here is too long, move it to the line above instead.

alicjakwie replied:
done

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+		return fmt.Errorf("CreateFile(id: %d).%v: could not create file %s: %w", fileID, status, fileName, err)
+	}
+	if err := wc.Close(); err != nil {
+		return fmt.Errorf("Writer.Close: %v", err)
```
Should we also add a status in this error as well?

alicjakwie replied:
I don't think there's one that could match it

deniadelcid replied:
I think this is an important error to record because it means that we weren't able to close the writer. Which could mean that there is a problem with the GCS bucket/file.

- [x] We should probably add a new error type here and then create a new PR to add it to the list of errors in metrics.go.

- [ ] You also may want to log this in the logger.

alicjakwie replied:
Made the changes in create_file.go all tests pass (with local changes to metrics.go) @evanSpendlove is opening a new pull request for adding this exit status to the metrics

deniadelcid commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	return fmt.Sprintf("Hermes_%02d_%v", f.ID, fmt.Sprintf("%x", checksum)), nil
+}
+
+func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSize int, client stiface.Client, logger *logger.Logger) error {
```
Just a suggestion, but you could also return the status of the error, i.e. (error, m.ExitStatus), this way on your test you could compare the exit status instead of an error message in scenarios where error is not nil.

You could also make a struct  (e.g. ProbeError) that contains both an  `error` and an `m.ExitStatus`) - probably metrics.go would be the place for it(?). And have CreateFile just return a ProbeError instead. 

You can/should leave this for a future PR thou, just open an issue and leave a TODO in monitor.go.

alicjakwie replied:
This will be left as a TODO in probe.go

deniadelcid replied:
I think this aligns with the requests from issue 16 [1].

[1] https://github.com/googleinterns/step224-2020/issues/16

deniadelcid commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+	tests := []struct {
+		file     RandomHermesFile
+		wantName string
+		wantErr  string
```
With the idea of having a `ProbeError` as the return value for CreateFile, you could change this line to `wantStatus  metrics.ProbeError`. Then your tests instead of having the error string could have something like `metrics.OutOfRange`, `metrics.Success`, or `metrics.InvalidArgument`

alicjakwie replied:
This will be left as a TODO in probe.go

deniadelcid commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+	"fmt"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	metricpb "github.com/google/cloudprober/metrics/proto"
+	probepb "github.com/googleinterns/step224-2020/config/proto"
+	"github.com/googleinterns/step224-2020/hermes/probe"
+	"github.com/googleinterns/step224-2020/hermes/probe/metrics"
+	"github.com/googleinterns/step224-2020/hermes/probe/mock"
+	journalpb "github.com/googleinterns/step224-2020/hermes/proto"
```
Can you reorganize these? The renamed imports should be at the bottom within their own block (i.e. leave a blank line between the renamed imports and those that haven't been renamed). I think that running `gofmt` should do this for you.

alicjakwie replied:
reorganised

deniadelcid commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+	}
+}
+
+func TestChecksum(t *testing.T) {
+	f := RandomHermesFile{11, 100}
+	fTwo := RandomHermesFile{13, 1000}
+	fCopy := RandomHermesFile{11, 100}
```
Can you also make a table for this test?

alicjakwie replied:
Restructured as discussed in our POD meeting

deniadelcid commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+		BucketName:             "test_bucket_probe0",
+	}
+
+	err := fmt.Errorf("") //err cannot be defined in the if statement below using ':=' as then the set up fails with the error:  expected identifier on left side of :=
```
Add a space after the `//` in your comment.

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func TestChecksum(t *testing.T) {
+	f := RandomHermesFile{11, 100}
```
I would structure this test as a table-driven test:

testCases := []struct{
   f *RandomHermesFile,
   want string
}{
  {&RandomHermesFile{11, 100}, "0xaa ...."},
  {&RandomHermesFile{13, 1000}, "0xaa ...."},
...
}

alicjakwie replied:
@deniadelcid raised the same issue https://github.com/googleinterns/step224-2020/pull/57#discussion_r489391429 and we discussed it in our POD meeting. Changed the test as per what was suggested -  it's not table driven but should hopefully be more readable than previously

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+		BucketName:             "test_bucket_probe0",
+	}
+
+	err := fmt.Errorf("") //err cannot be defined in the if statement below using ':=' as then the set up fails with the error:  expected identifier on left side of :=
```
You want `var err error` instead. Also, remove the comment.

alicjakwie replied:
changed and removed!

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	err := fmt.Errorf("") //err cannot be defined in the if statement below using ':=' as then the set up fails with the error:  expected identifier on left side of :=
+	if target.LatencyMetrics, err = metrics.NewMetrics(hp, probeTarget); err != nil {
```
If I understand it, target.LatencyMetrics is not used. If that's the case, then use the `_` placeholder to throw away that return value. Then you can skip the `var err error` part and use `:=` instead: `if _, err := ...; err != nil {`.

alicjakwie replied:
Maybe I misunderstood it but I think that here metrics are initialised in the LatencyMetrics field in target (which is a Target struct - defined in package probe). Is this right @evanSpendlove ? That's why I couldn't use ':=@ and had to define error beforehand

evanSpendlove replied:
target.LatencyMetrics is normally initialised inside the probe.go file. However, this is a test environment that doesn't use probe.go.

target.LatencyMetrics is used in create_file.go so it needs to be initialised before target is passed to CreateFile().

evanSpendlove replied:
E.g. line 128 of create_file.go

leczb replied:
I see! This is a set-up operation.
In this case use t.Fatalf() on failure.

alicjakwie replied:
Changed it to t.Fatalf()

alicjakwie committed:
Update create_file_test.go

changed wantErr to bool
changed TestChecksum

alicjakwie committed:
Update create_file_test.go

Now updating Intent
Added a check that a file actually gets created

alicjakwie committed:
Update create_file_test.go

deniadelcid commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+
+func TestChecksum(t *testing.T) {
+	file := RandomFile{11, 100}
+	otherFile := RandomFile{13, 1000}
```
Move this closer to where it's used. Also, at that time you can just re-use the "file" variable.

alicjakwie replied:
Done!

deniadelcid replied:
Can we move this line in between lines 61 and 62?

My reasoning is that when actual CreateFile calls are made, first a `RandomFile` will be created (like you do in lines 56 and 57) and immediately after you create the checksum for each (before creating the next one). So if I create two files (with IDs 1 and 2) the order of these two operations would be: 
1.  Create RandomFile{1, size}
2. Checksum for file_id = 1
3. Create RandomFile{2, size}
4. Checksum file with ID = 2

alicjakwie commented on hermes/probe/create/create_file_test.go:
```diff
@@ -5,149 +5,136 @@ package create
-		logger.Infof("Object %v added in bucket %s.", fileName, bucketName)
+	var err error
+	if target.LatencyMetrics, err = metrics.NewMetrics(hp, probeTarget); err != nil {
+		t.Error(err)
```
```suggestion
		t.Fatalf("Metric set up failed due to %s", err.Error())
```

alicjakwie committed:
Update hermes/probe/create/create_file_test.go

cianbgoog commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,130 @@
+	maxFileSizeBytes = 1000 // maximum allowed file size in bytes
+)
+
+type RandomHermesFile struct {
```
I'm just call this `GeneratedFile` - you don't need hermes, because it comes from Hermes, and users don't need to know that it's randomly generated.

alicjakwie replied:
called it RandomFile struct!

alicjakwie committed:
Delete create_file.go

alicjakwie committed:
Delete create_file_test.go

cianbgoog commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,140 @@
+		got, err := test.file.FileName()
+		if err == nil {
+			if test.wantErr {
+				t.Errorf("{%v, %v}.FileName() failed expected an error got nil", test.file.ID, test.file.Size)
```
nit: You should write this as if it was go code so `fmt.Errorf("test.file{%v}.FileName() = %v wanted %v, test.file, got, wantName`
You can do something similar for tests where you want an error (or don't want one, and expected one).

This should ideally mean you have less branches here.

alicjakwie replied:
Will add that as a TODO changing error types and messages

alicjakwie replied:
Added

alicjakwie replied:
(#72 )

alicjakwie replied:
changed the error messages!

cianbgoog reviewed (approved):
minor changes, happy for you to merge when they're made.

alicjakwie committed:
addressed all the comments from the PR

alicjakwie committed:
Update create_file_test.go

alicjakwie committed:
Update create_file.go

alicjakwie committed:
Update create_file_test.go

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,160 @@
+		if got.ID != test.wantRF.ID || got.Size != test.wantRF.Size {
+			t.Errorf("{%v, %v}.NewRandomFile() failed expected {%v,%v}, got {%v,%v}", test.fileID, test.fileSize, test.wantRF.ID, test.wantRF.Size, got.ID, got.Size)
+		}
+
```
Nit: drop empty line.

alicjakwie replied:
Done!

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,160 @@
+	for _, test := range tests {
+		got, err := NewRandomFile(test.fileID, test.fileSize)
+		if err != nil && !test.wantErr {
+			t.Errorf("{%v, %v}.NewRandomFile() failed and returned an unexpected error %s", test.fileID, test.fileSize, err.Error())
```
Use the "%v" formatting verb and pass in `err`. The formatting code will automatically call the `Error()` method on `err`.

alicjakwie replied:
Changed!

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,160 @@
+		{3, 0, true, &RandomFile{}},
+		{3, 1001, true, &RandomFile{}},
+	}
+	for _, test := range tests {
```
Nit: The loop variable is typically called `tc` (for test-case).
In general, use short (typically single-letter) names for the loop variable.

alicjakwie replied:
Changed!

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,160 @@
+		fileID   int64
+		fileSize int
+		wantErr  bool
+		wantRF   *RandomFile
```
Nit: this can be called `want`.
Also, re-order this to match the return order: `want` first then `wantErr`.

alicjakwie replied:
Changed!

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,160 @@
+		if err != nil {
+			t.Errorf("{%v, %v}.FileName() failed and returned an unexpected error %s", test.file.ID, test.file.Size, err.Error())
+		}
+		if got[0:9] != test.wantName {
```
9 is a magic number.

alicjakwie replied:
Constant added!

leczb commented on hermes/probe/create/create_file_test.go:
```diff
@@ -0,0 +1,160 @@
+
+	var err error
+	if target.LatencyMetrics, err = metrics.NewMetrics(hp, probeTarget); err != nil {
+		t.Fatalf("Metric set up failed due to %s", err.Error())
```
Nit: Just t.Fatalf("metrics.NewMetrics(): %v", err).
In general: separate the context information from the underlying error with `: `.

alicjakwie replied:
changed!

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	"github.com/googleapis/google-cloud-go-testing/storage/stiface"
+	"github.com/googleinterns/step224-2020/hermes/probe"
+
+	m "github.com/googleinterns/step224-2020/hermes/probe/metrics"
```
Don't rename imports for convenience.

alicjakwie replied:
removed!

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	"github.com/googleinterns/step224-2020/hermes/probe"
+
+	m "github.com/googleinterns/step224-2020/hermes/probe/metrics"
+	journalpb "github.com/googleinterns/step224-2020/hermes/proto"
```
Optional: Just `pb`, as this is the only proto import.

alicjakwie replied:
changed!

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+)
+
+type RandomFile struct {
+	ID   int64 // ID is a positive integer in the range [1,50]
```
Nit: Move comments above the fields.

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+
+type RandomFile struct {
+	ID   int64 // ID is a positive integer in the range [1,50]
+	Size int   // File size in bytes
```
The GoDoc comment should start with the identifier: "Size is the file size in bytes."
Also, comments should be full sentences, terminated with a period.

alicjakwie replied:
changed

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	maxFileSizeBytes = 1000 // maximum allowed file size in bytes
+)
+
+type RandomFile struct {
```
Exported identifiers must be documented.
Fix throughout.

alicjakwie replied:
documentation added!

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	return n, err
+}
+
+// Warning: NewReader is not thread safe
```
Nit: NewReader() is thread-safe. The Reader instance it returns is not.

alicjakwie replied:
Changed the GoDoc!

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+
+func NewRandomFile(fileID int64, fileSize int) (*RandomFile, error) {
+	if fileID < begin || fileID > end {
+		return &RandomFile{}, fmt.Errorf("the file ID provided %v wasn't in the required range [%v, %v]", fileID, begin, end)
```
`return nil, fmt.Errorf(...`
Here and below.

alicjakwie replied:
changed and changes applied in the test file to match

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	if fileSize > maxFileSizeBytes || fileSize <= 0 {
+		return &RandomFile{}, fmt.Errorf("invalid argument: RandomFile.Size = %d; want 0 < size <= %d", fileSize, maxFileSizeBytes)
+	}
+	return &RandomFile{ID: int64(fileID), Size: fileSize}, nil
```
Why the type conversion: can't we use `int` everywhere for the file ID?

alicjakwie replied:
Unfortunately not because rand.NewSource() requires an int64 as an argument added the casting to int64 there and added this comment. The argument to CreateFile is an int32 because the State Journal contains a map [int32] string.

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	return fmt.Sprintf("Hermes_%02d_%x", f.ID, checksum), nil
+}
+
+func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSize int, client stiface.Client, logger *logger.Logger) error {
```
The `RandomFile.ID` field is declared as `int64`, but we expect `int32` here. Is it intentional?
If not, change everything to `int`.

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	target.Journal.Intent = &journalpb.Intent{FileOperation: journalpb.Intent_CREATE, Filename: fileName}
+	if _, ok := target.Journal.Filenames[fileID]; ok {
+		var status m.ExitStatus
```
There's no need for this intermediate variable (even if you need it, just use `:=`).
Just
```
return fmt.Errorf("...", fileID, metrics.UnknownFileFound)
```

At the same time, this is a lossy conversion, as we convert `metrics.UnknownFileFound` into a string.
It would be better to include it as a wrapped error or a field on a custom error type.

alicjakwie replied:
The variable is now declared earlier in code and used throughout the whole function

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+		default:
+			status = m.ProbeFailed
+		}
+		target.LatencyMetrics.APICallLatency[m.APICreateFile][status].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
```
"hermes_api_latency_s" is a string repeated throughout the code. It should be a string const, otherwise it's very easy to make a typo that breaks the system in a subtle way.

alicjakwie replied:
Added a constant hermesAPILatency = "hermes_api_latency_s"

alicjakwie replied:
sorry hermesAPILatencySeconds*

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+		return fmt.Errorf("Writer.Close: %v", err)
+	}
+	target.LatencyMetrics.APICallLatency[m.APICreateFile][m.Success].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
+	prefix := fileName[0:9]
```
9 is a magic number.

alicjakwie replied:
added a constant

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	target.LatencyMetrics.APICallLatency[m.APICreateFile][m.Success].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
+	prefix := fileName[0:9]
+	query := &storage.Query{Prefix: prefix}
```
I don't think there's a need for the `prefix` intermediate variable.

alicjakwie replied:
removed

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+		return fmt.Errorf("CreateFile check failed: %w", err)
+	}
+	if obj.Name != fileName {
+		fmt.Errorf("CreateFile check failed expected file name present %v got %v", fileName, obj.Name)
```
Nit: Use the "%q" formatting verb for strings. It add quotes automatically.
I would recommend fixing this throughout the code for consistency.
It's good practice to use quotes around strings in error and log messages as it makes it easy to detect silly mistakes like empty string or extra leading/trailing spaces.

alicjakwie replied:
changed to %q

leczb commented on hermes/probe/create/create_file.go:
```diff
@@ -0,0 +1,160 @@
+	target.LatencyMetrics.APICallLatency[m.APIListFiles][m.Success].Metric("hermes_api_latency_s").AddFloat64(end.Sub(start).Seconds())
+
+	target.Journal.Filenames[fileID] = fileName
+	if logger != nil {
```
I would put a contract in the docstring the caller must pass in a valid logger. Then there's no need for these checks.

alicjakwie replied:
In the test file the logger passed is nil because w don't want to bring up cloudprober hence the check in the code. Adding doc strings is left as a TODO

leczb replied:
You could pass in a fake logger that just calls fmt.Println().

alicjakwie replied:
replaced! added the fakeLogger interface to the fakegcs file and package for now but do you think it should be its own separate package perhaps fakelogger?

leczb replied:
I think it's cleaner to put it in a separate package.

cianbgoog commented on hermes/probe/create/create_file.go:
```diff
@@ -21,91 +27,99 @@ const (
+func (f *RandomFile) FileName() (string, error) {
 	checksum, err := f.CheckSum()
 	if err != nil {
 		return "", err
```
You should anote this error - I would be somewhat surprised if Filename() failed because of an error with `io.copy`

Also, does FileName need to be exported?

alicjakwie replied:
It does not need to be exported changed to a non-exported function! checksum as well

alicjakwie replied:
Errors wrapped with %w now

alicjakwie replied:
apart from Fatalf because there %v is needed

cianbgoog commented on hermes/probe/create/create_file.go:
```diff
@@ -119,12 +133,28 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
 		return fmt.Errorf("CreateFile(id: %d).%v: could not create file %s: %w", fileID, status, fileName, err)
 	}
 	if err := wc.Close(); err != nil {
+		target.LatencyMetrics.APICallLatency[m.APICreateFile][m.WriterCloseFailed].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
```
`"hermes_api_latency_seconds"`, here & elsewhere

alicjakwie replied:
"hermes_api_latency_s" is defined in metrics.go I did as per @leczb suggestion add a constant hermesAPILatencySeconds="hermes_api_latency_seconds"

cianbgoog replied:
If we're defining latency names in metrics.go, can we just define the metric names as exported constants there, rather than having them in several places (and things breaking if they stop matching)

cianbgoog reviewed (approved):
minor comments, feel free to merge once fixed.

alicjakwie committed:
Update and rename create_file.go to create.go

alicjakwie committed:
Update and rename create_file_test.go to create_test.go

alicjakwie committed:
Update create.go

alicjakwie committed:
Update create_test.go

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+//
+// Author: Alicja Kwiecinska, GitHub: alicjakwie
+//
+// create contains all of the logic necessary to create the files with randomized contents in GCS
```
Nit: It should be "Package create ...".

alicjakwie replied:
changed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+)
+
+const (
+	FileNamePrefixSize      = 9    // len("Hermes_ID")
```
Nit: I would call it "...Length" instead of "...Size".

You can promote the comment to code:
`const FileNamePrefixSize = len("Hermes_ID")`

Even better:
```
const FileNameFormat = "Hermes_%02d_%x"
var  FileNamePrefixLength = len(fmt.Sprintf(FileNameFormat, 0, ""))
```

Note that this requires the FileNamePrefixLength to be declared as a variable, but it's totally acceptable.

alicjakwie replied:
should the variable be declared in the create function and exported or should it be declared outside of the function body?

leczb replied:
If the length is required elsewhere, make it a package-level variable.
If it's only required in Create(), then move it there (closest to its usage).

alicjakwie replied:
we need it in create_test.go

alicjakwie replied:
Acc (caught this in my tests) var  FileNamePrefixLength = len(fmt.Sprintf(FileNameFormat, 0, ""))-1 as this prefix has two __ so it's of the form Hermes_00_ instead of Hermes_00

alicjakwie replied:
acc we only need FileNamePrefixLength in create_test so only added it there and made it non-exported

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+
+const (
+	FileNamePrefixSize      = 9    // len("Hermes_ID")
+	begin                   = 1    // begin is the ID of the first HermesFile.
```
`minFileID` would be self-describing, obviating the need for the comment.

alicjakwie replied:
changed comment removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+const (
+	FileNamePrefixSize      = 9    // len("Hermes_ID")
+	begin                   = 1    // begin is the ID of the first HermesFile.
+	end                     = 50   // end is the ID of the last HermesFile.
```
`maxFileID`

alicjakwie replied:
changed comment removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+	begin                   = 1    // begin is the ID of the first HermesFile.
+	end                     = 50   // end is the ID of the last HermesFile.
+	maxFileSizeBytes        = 1000 // maxFileSizeBytes is the maximum allowed file size in bytes.
+	hermesAPILatencySeconds = "hermes_api_latency_s"
```
This should be in the metrics package.
Sync with Evan about this.

alicjakwie replied:
@evanSpendlove

evanSpendlove replied:
It's in the latest commit to the open metrics PR.

It doesn't need to be changed here. I will change it in the metrics PR after this is PR is merged.

alicjakwie replied:
Thank you Evan!

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+)
+
+type randomFile struct {
+	id   int32 // id is a positive integer in the range [begin,end]
```
Why not `int`?

alicjakwie replied:
added a TODO(#76)

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+
+type randomFile struct {
+	id   int32 // id is a positive integer in the range [begin,end]
+	size int   // size is the file size in bytes.
```
If you call this `sizeBytes`, then the comment can be removed.

alicjakwie replied:
replaced comments removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+	rand *rand.Rand
+}
+
+// Read implements the standard Read interface, it reads data from a randomFileReader.
```
Nit: It doesn't read data from `randomFileReader`. It returns a random stream of bytes generated by a pseudo-random  generator.

alicjakwie replied:
changed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+
+// Read implements the standard Read interface, it reads data from a randomFileReader.
+// Argument:
+//      buf: a byte slice serves as a buffer
```
"as an output buffer"

alicjakwie replied:
changed!

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+// Argument:
+//      buf: a byte slice serves as a buffer
+// Returns:
+//	n:  the size of the buffer buf following a Read operation
```
`n` is the number of bytes read.
The buffer's size doesn't change.

alicjakwie replied:
changed!

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+	return n, err
+}
+
+// NewReader creates a new randomFileReader and returns a pointer to it
```
Nit: There's no need to repeat type-related information in the docstring. I would simplify the comment to
"NewReader returns an io.Reader instance."

alicjakwie replied:
changed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+// Returns:
+//	*randomFileReader: a Reader that supports randomFiles. This Reader contains information about the file size in bytes as well as a pseudo-random bytes generator
+// Warning: The randomFileReader produced by NewReader is not thread safe.
+func (f *randomFile) NewReader() *randomFileReader {
```
Assuming this is only called in this package, make this non-exported.

alicjakwie replied:
changed to non-exported doc strings removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+func (f *randomFile) fileName() (string, error) {
+	checksum, err := f.checksum()
+	if err != nil {
+		return "", fmt.Errorf("{%d, %d}.checksum = nil,  %v", f.id, f.size, err)
```
How about

`fmt.Errorf("failed to generate file checksum: %v", err)`

alicjakwie replied:
@cianbgoog suggested changing the error messages to match the go syntax does this not apply in this case?

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+	if err != nil {
+		return "", fmt.Errorf("{%d, %d}.checksum = nil,  %v", f.id, f.size, err)
+	}
+	return fmt.Sprintf("Hermes_%02d_%x", f.id, checksum), nil
```
Make the file prefix a const.

alicjakwie replied:
as in "Hermes_" because "Hermes_02%d" depends on the file id

leczb replied:
See my proposal above about introducing a FileNameFormat const.
You can use that const here.

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+	//  id will serve as a Seed and i - index of the currently read byte  will be set to 0 automatically in the returned reader
+	return &randomFileReader{
+		size: f.size,
+		rand: rand.New(rand.NewSource(int64(f.id))), // rand.NewSource() requires an int64 as an argument
```
Remove the comment. Go doesn't do automatic type conversion. If the reader sees this type conversion, it's obvious that `int64` is the expected type (otherwise it wouldn't compile).

alicjakwie replied:
comment removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+
+func newRandomFile(fileID int32, fileSize int) (*randomFile, error) {
+	if fileID < begin || fileID > end {
+		return nil, fmt.Errorf("the file id provided %d wasn't in the required range [%d, %d]", fileID, begin, end)
```
Nit: "the file ID".

alicjakwie replied:
I had ID capitalised but randomFile has a field: id (as we don't export it) so I am wondering which one to use in the comment?

leczb replied:
If you are referring to the `fileID` argument, then use "fileID" instead of "file id" or "file ID".
You can make the error message more succinct:
`fmt.Errorf("invalid argument: fileID = %d; want %d <= fileID <= %d", ...)`

alicjakwie replied:
changed!

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+func newRandomFile(fileID int32, fileSize int) (*randomFile, error) {
```
Why `int32`?

alicjakwie replied:
I think I mentioned this in a reply to a previous comment (if not sorry) but it's because in State Journal which we're updating FileNames is a map[int32] string

leczb replied:
Sure, but why is *that* int32 in turn?
Is it because it's a proto field? If so, make it int64 and convert at the last moment. Use `int` everywhere in the code.

alicjakwie replied:
@evanSpendlove made it map[int32] string so save space because it will be a max of 50 different types.
Should then fileID be an int and we should cast it to int32 and int64 in respective places?

leczb replied:
Using `int32` instead of `int` (which is typically `int64` on modern systems) doesn't save much space, especially if there are only that few entries in the map. It's a premature optimization. Just use `int` everwhere.

deniadelcid replied:
Agreed. But since for the StateJournal is in a separate file, I've opened issue to fix this: https://github.com/googleinterns/step224-2020/issues/76.

@alicjakwie : can you add a TODO on this file to fix this after create.go is merged?

alicjakwie replied:
added a TODO to change this (#76 )

deniadelcid commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+//
+// Author: Alicja Kwiecinska, GitHub: alicjakwie
+//
+// create contains all of the logic necessary to create the files with randomized contents in GCS
```
uber nit.
```suggestion
// create contains all of the logic necessary to create files with randomized contents in GCS.
```

leczb commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+		target.LatencyMetrics.APICallLatency[metrics.APIListFiles][metrics.FileMissing].Metric(hermesAPILatencySeconds).AddFloat64(finish.Sub(start).Seconds())
+		return fmt.Errorf("CreateFile check failed: %v", err)
+	}
+	if obj.Name != fileName {
```
This assumes that only a single result is returned by the query.
What if there are multiple results?

alicjakwie replied:
we check whether there was a file with this ID in the logger at the beginning of the create function and also we check whether a file with this ID exists on the storage system by calling the Read probe and if we call create after deletion there is a check performed at the end of deletion to see whether a file with this ID does not exist on the storage sysem so essentially if there are multiple results it would mean that we created more than one file during this probe run?

Therefore should we consider the case with the multiple results and if so should this then be changed to 
for {
} with a break statement if obj.Next = nil? 

or if there are multiple results  it clearly means that something went wrong so just return an error?

leczb replied:
I would use a for loop and also error out if more than one file is found (listing all the files found in the error).

alicjakwie replied:
changed to a loop and errored out if a list is found and also returning the list of files found

deniadelcid commented on hermes/probe/create/create.go:
```diff
@@ -0,0 +1,186 @@
+//      buf: a byte slice serves as a buffer
+// Returns:
+//	n:  the size of the buffer buf following a Read operation
+//	err: error it should be nil when not done reading and io.EOF once the whole file contents have been read
```
nit: align the margins of your arguments with those of your returns.
```suggestion
// Argument:
//	buf: a byte slice serves as a buffer
// Returns:
//	n:  the size of the buffer buf following a Read operation
//	err: error it should be nil when not done reading and io.EOF once the whole file contents have been read
```

alicjakwie committed:
Update create.go

alicjakwie committed:
Update create_test.go

leczb commented on hermes/probe/create/create.go:
```diff
@@ -35,85 +38,84 @@ import (
-	if len(buf) > (r.size - r.i) {
-		// if the length of buffer is greater than the number of bytes left to read make the size of the buffer match the number of bytes left to read
-		b = buf[:(r.size - r.i)]
+	if len(buf) > (r.sizeBytes - r.i) {
```
Nit: there's no need for the parentheses.

alicjakwie replied:
parentheses removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -35,85 +38,84 @@ import (
-		b = buf[:(r.size - r.i)]
+	if len(buf) > (r.sizeBytes - r.i) {
+		// if the length of buffer is greater than the number of bytes left to read make the sizeBytes of the buffer match the number of bytes left to read
+		b = buf[:(r.sizeBytes - r.i)]
```
Nit: there's no need for the parentheses.

alicjakwie replied:
parantheses removed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -35,85 +38,84 @@ import (
 	}
 	if fileSize > maxFileSizeBytes || fileSize <= 0 {
-		return nil, fmt.Errorf("invalid argument: randomFile.size = %d; want 0 < size <= %d", fileSize, maxFileSizeBytes)
+		return nil, fmt.Errorf("invalid argument: randomFile.sizeBytes = %d; want 0 < sizeBytes <= %d", fileSize, maxFileSizeBytes)
```
Nit: Replace `randomFile.sizeBytes` with `fileSize`, as that's the name of the argument.

alicjakwie replied:
arguments changed to sizeBytes as per previous comment

leczb commented on hermes/probe/create/create.go:
```diff
@@ -35,85 +38,84 @@ import (
 	}
 }
 
 func newRandomFile(fileID int32, fileSize int) (*randomFile, error) {
```
Nit: The `file` prefix in the arguments are not necessary. `id` and `size` (or preferably `sizeBytes`) would do just as well, as the reader already knows that it's about the file.

alicjakwie replied:
changed

leczb commented on hermes/probe/create/create.go:
```diff
@@ -155,27 +157,38 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
-		return fmt.Errorf("CreateFile check failed: %v", err)
+		return fmt.Errorf("CreateFile check failed: %w", err)
+	}
+	namesFound := []string{obj.Name}
```
It's surprising that we initialize the slice with a single element. Why isn't this an empty slice?
`var namesFound []string`

alicjakwie replied:
I will change it to an empty slice and change the order of operations in the loop!

leczb commented on hermes/probe/create/create.go:
```diff
@@ -155,27 +157,38 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
 	start = time.Now()
 	objIter := client.Bucket(bucketName).Objects(ctx, query)
 	finish := time.Now()
 	obj, err := objIter.Next()
```
I think this should be removed.

alicjakwie replied:
changed to fileNamePrefix := fmt.Sprintf(FileNameFormat, fileID, "")
	query := &storage.Query{Prefix: fileNamePrefix}
as fileNamePrefix gets used again

leczb commented on hermes/probe/create/create.go:
```diff
@@ -155,27 +157,38 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
+	query := &storage.Query{Prefix: fmt.Sprintf(FileNameFormat, fileID, "")}
 	start = time.Now()
 	objIter := client.Bucket(bucketName).Objects(ctx, query)
 	finish := time.Now()
```
Move the finish to where it actually finishes.

alicjakwie replied:
This is the APILatency that's why the finish is here

leczb replied:
The API operation in this case is "looking up all filenames matching a query".
That includes iterating over all items, as each iteration might incur a round-trip to the service.
This means the API operation is only "done" when we the iteration is done.

alicjakwie replied:
Okay thank you for explaining this changed!

leczb commented on hermes/probe/create/create.go:
```diff
@@ -155,27 +157,38 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
+		obj, err = objIter.Next()
+	}
+	if len(namesFound) > 1 {
+		fmt.Errorf("Found more than one file on the target system with the fileID: %d found: %s", fileID, strings.Join(namesFound, "\n"))
```
Nit: error messages should start with lower-case.
Also, you can make the error message more succinct:
`fmt.Errorf("expected exactly one file in bucket %q with prefix %q; found %d: %v", bucketName, fmt.Sprintf(FileNameFormat, fileID, ""), len(namesFound), namesFound)`

I would also move `fmt.Sprintf(FileNameFormat, fileID, "")` into a variable to avoid the repetition.

alicjakwie replied:
created a varialble fileNamePrefix moved fmt.Sprintf(FileNameFormat, fileID, "")

leczb commented on hermes/probe/create/create.go:
```diff
@@ -155,27 +157,38 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
+		namesFound = append(namesFound, obj.Name)
+		obj, err = objIter.Next()
+	}
+	if len(namesFound) > 1 {
```
I would add another check: what if it's zero?
Either add another `if` statement or change the condition to `!= 1`.

alicjakwie replied:
That's caught in obj, err := objIter.Next()
	if err != nil {
		target.LatencyMetrics.APICallLatency[metrics.APIListFiles][metrics.FileMissing].Metric(hermesAPILatencySeconds).AddFloat64(finish.Sub(start).Seconds())
		return fmt.Errorf("CreateFile check failed: %w", err)
	}

leczb replied:
I would simplify it like this:
1. Execute query
2. Get all results with a for loop
3. Check that we got exactly one result

alicjakwie replied:
and the exit status is FileMissing then which is not equivalent to the error given when too many files are found

alicjakwie replied:
I changed the order now but I seperated the case no result and too many results to include the status file missing

leczb commented on hermes/probe/create/create_test.go:
```diff
@@ -72,18 +74,18 @@ func TestFileName(t *testing.T) {
 		}
-		if got[:FileNamePrefixSize] != tc.want {
-			t.Errorf("{%d, %d}.fileName() =  %q_checksum expected %q_checksum", tc.file.id, tc.file.size, got[:FileNamePrefixSize], tc.want)
+		if got[:fileNamePrefixLength] != tc.want {
```
Use `strings.HasPrefix()` instead.

alicjakwie replied:
changed

alicjakwie committed:
Update create.go

alicjakwie committed:
Update create_test.go

alicjakwie committed:
Update create.go

leczb commented on hermes/probe/create/create.go:
```diff
@@ -168,25 +167,26 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
+	finish := time.Now()
+	if len(namesFound) == 0 {
+		target.LatencyMetrics.APICallLatency[metrics.APIListFiles][metrics.FileMissing].Metric(hermesAPILatencySeconds).AddFloat64(finish.Sub(start).Seconds())
+		return fmt.Errorf("CreateFile check failed: %w", err)
```
There's no underlying error here.
What fails here is that we expected exactly one result from the query but got none.
If there was an unrelated error (let's say a network issue), it should have been caught in the loop above.

alicjakwie replied:
Maybe I'm misunderstanding sorry but isn't the difference the fact that if there are 0 files then for sure our create file operation must have failed and not create a file and if there is more than one file it could be another issue?

leczb replied:
There are 4 different outcomes
1) The query itself fails due to some error
2) We get zero results
3) We get one result
4) We get more than one result

#1 is not handled by the current code.
I would re-arrange the code to follow the pattern in n https://godoc.org/cloud.google.com/go/storage#hdr-Listing_objects.
Then a query failure will be handled.

alicjakwie replied:
I will I just wonder whether we should distinguish between 2 and 4

leczb replied:
We must distinguish #2 and #4: if we get zero results here that's a signal that creating the file failed silently (we didn't get any error when we finished writing the file on line 161.)
If we get one result (and the filename matches our expectation), the creation succeeded.

alicjakwie replied:
Distinguishing between #2 and #4 now

leczb commented on hermes/probe/create/create.go:
```diff
@@ -168,25 +167,26 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
 	start = time.Now()
 	objIter := client.Bucket(bucketName).Objects(ctx, query)
-	finish := time.Now()
 	obj, err := objIter.Next()
```
Remove this and make it the first action in the for loop.
Follow the pattern in https://godoc.org/cloud.google.com/go/storage#hdr-Listing_objects.

alicjakwie replied:
noted

alicjakwie replied:
changed

alicjakwie committed:
Update create.go

leczb commented on hermes/probe/create/create.go:
```diff
@@ -171,25 +172,27 @@ func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSiz
 	}
 	if namesFound[0] != fileName {
-		fmt.Errorf("CreateFile check failed expected file name present %q got %q", fileName, obj.Name)
+		return fmt.Errorf("CreateFile check failed expected file name present %q got %q", fileName, namesFound[0])
```
Nit: Make it more succinct:
`fmt.Errorf("CreateFile check failed: filename matching %q prefix: %q; want %q", fileNamePrefix, namesFound[0], filaName)`

alicjakwie replied:
changed!

leczb reviewed (approved):
LGTM

evanSpendlove reviewed (approved):
LGTM

alicjakwie committed:
Update create.go
